Implement `odgi` package

tbhaxor opened the pull request:
This PR adds CMakePackage odgi package and all its available releases as of now.

> **About the package**
> 
> Optimized Dynamic Genome/Graph Implementation: understanding pangenome graphs. `odgi` provides an efficient and succinct dynamic DNA sequence graph model, as well as a host of algorithms that allow the use of such graphs in bioinformatic analyses.

tbhaxor committed:
feature (packages): add cmake package odgi
This PR adds all available versions of odgi package.

tbhaxor committed:
fix (packages): lint and format odgi

This PR simply fixes the lint error reported by the CI job
[here](https://github.com/spack/spack/actions/runs/5467496550/jobs/9953889878?pr=38741#step:6:67)
for `odgi` package.

becker33 commented on var/spack/repos/builtin/packages/odgi/package.py:
```diff
@@ -0,0 +1,44 @@
+
+    # dependencies
+    depends_on("libdivsufsort", when="+libdivsufsort")
+    depends_on("sdsl-lite", when="+sdsl-lite +static")
```
If the `sdsl-lite` variant doesn't do anything when `~static`, we should write a conflict

```
conflicts("+sdsl-lite~static")
```

tbhaxor replied:
Its the case only when ninja is used as generator

wdconinc committed:
mlpack: new package (#38277)

* mlpack: new package

mlpack is an intuitive, fast, and flexible header-only C++ machine learning library with bindings to other languages. It is meant to be a machine learning analog to LAPACK, and aims to implement a wide array of machine learning methods and functions as a "swiss army knife" for machine learning researchers.

* mlpack: upstream merged patch to allow python installation in spack

EbiArnie reviewed (changes requested):
I would recommend a few changes.

- I have not been able to build this as is. To build, this software needs git submodules. The downloadable tarballs don't include these and the cmake step can not download them.
I based the package on a git checkout with submodules=True

- I don't think having variants for sdsl-lite and libdivsufsort makes too much sense.
odgi needs them and builds them if not provided. Actually, right now, the pkgconfig for sdsl-lite is wrong and only includes libdivsufsort. This results in sdsl-lite always being built.

- I have included python and pybind11. This is not optional. If we don't provide this, the system python is picked up instead. We don't want this. For pybind11 we have to provide a CMAKE parameter so that the correct python and config is found.

- It is not necessary to depend on python+debug, normal python is enough.

- There are dependencies on openmpi.


- The bits for a static build can stay of course.


This works for me:



```

class Odgi(CMakePackage):   
    git = "https://github.com/pangenome/odgi.git"
    version("0.8.3", commit="34f006f31c3f6b35a1eb8d58a4edb1c458583de3", submodules=True)
    conflicts("%gcc@:9.3")
    depends_on("python")
    depends_on("py-pybind11")
    depends_on("openmpi")
    depends_on("sdsl-lite")
    depends_on("libdivsufsort")
    depends_on("jemalloc")

    def cmake_args(self):
        args = []
        args.append("-DPYTHON_EXECUTABLE:FILEPATH={0}".format(self.spec["python"].command))
        return args

```

tbhaxor commented:
Thanks for your review. I will work on this today and fix it

wdconinc committed:
qt-* (Qt6 pkgs): new versions 6.5.0, 6.5.1 (#36705)

* qt-base: new version 6.5.0

* qt-declarative: new version 6.5.0

* qt-quick3d: new version 6.5.0

* qt-quicktimeline: new version 6.5.0

* qt-shadertools: new version 6.5.0

* qt-*: new version 6.5.1

* qt-base: new version 6.5.1

tbhaxor committed:
improve (packages): add odgi git package

This PR removes odgi package created from the git release and introduces
odgi-git to source the package from git repository. However this commit
only implements v0.8.3 at 34f006f31c3f6b35a1eb8d58a4edb1c458583de3 along
with submodules.

Thanks to @EbiArnie

tbhaxor committed:
improve (lib): type hints for self.spec

This PR adds the type hints to self.spec member of
`spack.package_base.PackageBase` class.

mtaillefumier committed:
Update cp2k recipe to use cmake or the current build system (#35718)

* Update cp2k recipe to use cmake or the current build system

Offers the possibility to build cp2k with the new cmake build system. commands like this are now supported

spack install cp2k@master build_system=cmake +.....

the recipe supports the following optional functionalities

- superlu, cosma, sirius, spglib, metis, spglib, libxc, libint, cuda/rocm, mkl/openblas/sci (and others), mpi, openmp, dbcsr
- dbcsr is built separately using the currently available recipe.

Two PRs need to be merged to be fully functional (cosma update in spack + one PR in cp2k github).

* Fix indentation

* Fix indentation

* Update libvori

* More typos

* Simplify BLAS/LAPACK

* Simplify BLAS/LAPACK

* Add A100 gpu value

* Fix typo

* Add the enable_regtests option

if -DCP2K_ENABLE_REGTESTS=ON (+enable_regtests with spack) then the location of the binary executables will be in the cp2k root directory under exe/build-cmake-*. This option is needed to run the regtests afterwards.

* Minor update

* more fixes

* Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Mikael Simberg <mikael.simberg@iki.fi>

* Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Mikael Simberg <mikael.simberg@iki.fi>

* Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Mikael Simberg <mikael.simberg@iki.fi>

* Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Mikael Simberg <mikael.simberg@iki.fi>

* Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Mikael Simberg <mikael.simberg@iki.fi>

* Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Mikael Simberg <mikael.simberg@iki.fi>

* small changes

* Remove any reference to nvidia architecture in the rocm list

* Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Alberto Invernizzi <9337627+albestro@users.noreply.github.com>

* Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Alberto Invernizzi <9337627+albestro@users.noreply.github.com>

* Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Alberto Invernizzi <9337627+albestro@users.noreply.github.com>

* Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Alberto Invernizzi <9337627+albestro@users.noreply.github.com>

* Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Alberto Invernizzi <9337627+albestro@users.noreply.github.com>

* Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Alberto Invernizzi <9337627+albestro@users.noreply.github.com>

* Final reformating

* Update py-fypp

---------

Co-authored-by: Mikael Simberg <mikael.simberg@iki.fi>
Co-authored-by: Alberto Invernizzi <9337627+albestro@users.noreply.github.com>

wdconinc committed:
mlpack: remove go variant from cmake_args (#38821)

michaelkuhn committed:
installer: Improve status reporting (#37903)

Refactor `TermTitle` into `InstallStatus` and use it to show progress
information both in the terminal title as well as inline. This also
turns on the terminal title status by default.

The inline output will look like the following after this change:
```
==> Installing m4-1.4.19-w2fxrpuz64zdq63woprqfxxzc3tzu7p3 [4/4]
```

rbberger committed:
legion: add ofi-slingshot11 conduit (#38859)

vicentebolea committed:
vtk-m: correct cuda_arch variant behavior (#38697)


Co-authored-by: eugeneswalker <eugenesunsetwalker@gmail.com>

vicentebolea committed:
adios2: add aws variant (#38857)

dependabot[bot] committed:
build(deps): bump docker/setup-buildx-action from 2.8.0 to 2.9.0 (#38783)

Bumps [docker/setup-buildx-action](https://github.com/docker/setup-buildx-action) from 2.8.0 to 2.9.0.
- [Release notes](https://github.com/docker/setup-buildx-action/releases)
- [Commits](https://github.com/docker/setup-buildx-action/compare/16c0bc4a6e6ada2cfd8afd41d22d95379cf7c32a...2a1a44ac4aa01993040736bd95bb470da1a38365)

---
updated-dependencies:
- dependency-name: docker/setup-buildx-action
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

vicentebolea committed:
vtk-m: modernize vtk-m recipe (#38726)

adamjstewart committed:
Standardize subcommand help strings (#38804)

### Rationale

While working on #29549, I noticed a lot of inconsistencies in our argparse help messages. This is important for fish where these help messages end up as descriptions in the tab completion menu. See https://github.com/spack/spack/pull/29549#issuecomment-1627596477 for some examples of longer or more stylized help messages.

### Implementation

This PR makes the following changes:

- [x] help messages start with a lowercase letter.
- [x] Help messages do not end with a period
- [x] the first line of a help message is short and simple

    longer text is separated by an empty line
- [x] "help messages do not use triple quotes" 

    """(except docstrings)"""
- [x] Parentheses not needed for string concatenation inside function call
- [x] Remove "..." "..." string concatenation leftover from black reformatting
- [x] Remove Sphinx argument docs from help messages

The first 2 choices aren't very controversial, and are designed to match the syntax of the `--help` flag automatically added by argparse. The 3rd choice is more up for debate, and is designed to match our package/module docstrings. The 4th choice is designed to avoid excessive newline characters and indentation. We may actually want to go even further and disallow docstrings altogether.

### Alternatives

Choice 3 in particular has a lot of alternatives. My goal is solely to ensure that fish tab completion looks reasonable. Alternatives include:

1. Get rid of long help messages, only allow short simple messages
2. Move longer help messages to epilog
3. Separate by 2 newline characters instead of 1
4. Separate by period instead of newline. First sentence goes into tab completion description

The number of commands with long help text is actually rather small, and is mostly relegated to `spack ci` and `spack buildcache`. So 1 isn't actually as ridiculous as it sounds.

Let me know if there are any other standardizations or alternatives you would like to suggest.

manuelakuhn committed:
Restrict py-pip version for packages using --install-option (#38837)

dependabot[bot] committed:
build(deps): bump docker/setup-buildx-action from 2.9.0 to 2.9.1 (#38868)

Bumps [docker/setup-buildx-action](https://github.com/docker/setup-buildx-action) from 2.9.0 to 2.9.1.
- [Release notes](https://github.com/docker/setup-buildx-action/releases)
- [Commits](https://github.com/docker/setup-buildx-action/compare/2a1a44ac4aa01993040736bd95bb470da1a38365...4c0219f9ac95b02789c1075625400b2acbff50b1)

---
updated-dependencies:
- dependency-name: docker/setup-buildx-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

haampie committed:
Add type hints to spack.installer (#38872)

haampie committed:
mirrors: distinguish between source/binary mirror; simplify schema (#34523)

Allow the following formats:

```yaml
mirrors:
  name: <url>
```

```yaml
mirrors:
  name:
    url: s3://xyz
    access_pair: [x, y]
```

```yaml
mirrors:
  name:
    fetch: http://xyz
    push:
      url: s3://xyz
      access_pair: [x, y]
```

And reserve two new properties to indicate the mirror type (e.g.
mirror.spack.io is a source mirror, not a binary cache)

```yaml
mirrors:
  spack-public:
    source: true
    binary: false
    url: https://mirror.spack.io
```

haampie committed:
spack buildcache push: improve argparse (#38876)

EbiArnie commented on var/spack/repos/builtin/packages/odgi-git/package.py:
```diff
@@ -0,0 +1,50 @@
+from spack.package import *
+
+
+class OdgiGit(CMakePackage):
```
I disagree that this should be called odgi-git. I would prefer just 'odgi', matching the tool and the git name.

I want to depend on this by the name 'odgi' or 'odgi@version'. Having an odgi and an odgi-git package means I have to be aware of the two packages and change the name when chaging from a git version to a non-git version. This can easily happen when odgi does a proper release version.

tbhaxor replied:
Just so you know, when I created a package from the git repository, it created a '-git' suffixed package.

tbhaxor replied:
I am doing as you said

EbiArnie commented on var/spack/repos/builtin/packages/odgi-git/package.py:
```diff
@@ -0,0 +1,50 @@
+
+    # the range is required to successfully build the program
+    conflicts(
+        "%gcc@:9.2,12.3.1:", msg="Unsupported compiler version. Recommended range is 9.3 -> 12.3"
```
I'd go with
"%gcc@:9.2,13:"
here. I did not test with gcc@13 though, so I take your word that this does not work.

tbhaxor replied:
I have tested with gcc@13.1 and it game compilation error, but works fine with 12.x latest version

EbiArnie commented on var/spack/repos/builtin/packages/odgi-git/package.py:
```diff
@@ -0,0 +1,50 @@
+    )
+
+    # <<< Dependencies list starts here
+    requires("%gcc@9.3:12", when="%gcc")
```
I recommend this here:
```
# Odgi needs gcc
    requires(
        "%gcc",
        msg="Package odgi depends on the gcc C++ compiler",
    )
```

tbhaxor replied:
This is also done

EbiArnie commented on var/spack/repos/builtin/packages/odgi-git/package.py:
```diff
@@ -0,0 +1,50 @@
+
+    def cmake_args(self):
+        args = [
+            "-DCMAKE_CXX_STANDARD:STRING=11",
```
The CMakeList for odgi specifies:
set(CMAKE_CXX_STANDARD 17)
I think this is not actually necessary (or even wrong)?
The 'cxx_standard_required' below seems necessary though.

tbhaxor replied:
Makes sense. It is done already

tbhaxor committed:
fix (packages): remove cxx standard in odgi-git

This PR removes `-DCMAKE_CXX_STANDARD:STRING=11` line from the cmake_args
function because it is already set in the CMakeLists.txt of the source
repository.

tbhaxor committed:
improve (packages): add requires condition in odgi-git

GCC is recommended to build the package. Therefore its been added before
gcc version conflicts statement.

tbhaxor committed:
chores (packages): rename odgi-git to odgi

This commit renames the odgi-git to odgi as asked [here](https://github.com/spack/spack/pull/38741#discussion_r1272265440)
